### ISSUE TITLE
Be able to specify path for Generated code

### ISSUE THREAD

**Issue by neelofer** [apollographql/apollo-android#2019]
**Is your feature request related to a problem? Please describe.**
I would like a way to provide the path for generated code for a service. 

**Describe the solution you'd like**
Right now the code is generated in `{project_dir/build/generated/source/apollo/main/{service name}/{rootPackagePath}` path. It would nice to specify a path separately, related to the project dir. E.g. something like `{project_dir}/src/main/generated/...`. This is to be able to merge generated code in Git, similar to Swagger type behavior.

What I am using:
gradle 6.2.1
'com.apollographql.apollo' version '1.3.2'
Java 8

My build.gradle task:
```
apollo {
    onCompilationUnit {
        outputDir.set(file("src/main/generated")) //trying to set this did not help me
    }
    generateKotlinModels.set(false)
    service("sample") {
        sourceFolder.set("com/sample/servicename/path")
        rootPackageName.set("com.sample.servicename.path")
    }
}
```

**Comment by martinbonnin** [comment]
Actually, i'd prefer swagger to generate files outside `src/main/kotlin`, see https://github.com/Yelp/swagger-gradle-codegen/pull/102#discussion_r381960028

I believe the generated code shouldn't be checked in source control because it duplicates information with the `.graphql` files. 

Can you detail a bit more what your use cases would be ?  Maybe there's a way to do it without adding files in the main source set ?

**Comment by martinbonnin** [comment]
Closing due to inactivity, please reopen if needed.

**Comment by tobiasdiez** [comment]
We have the same issue. Other uses cases are IDE integration (auto completion) which sometimes doesn't work / needs special configuration, and usage with the Java module system. You also don't necessarily have to have the code under version control, simply add `src/main/generated` (or whatever path you choose) to gitignore.

So I hope this issue could be reopened. Thanks!

**Comment by martinbonnin** [comment]
Hi @tobiasdiez !

**Comment by martinbonnin** [comment]
> IDE integration  (auto completion) which sometimes doesn't work / needs special configuration

If autocomplete doesn't work, it's definitely a bug that should be addressed either in Apollo Android or IntelliJ/Android Studio. Would you have a small reproducer we could dig into?

> usage with the Java module system

Can you elaborate a bit more what is specific to the Java module system? I was under the impression that the generated files would be part of the module where the apollo plugin is applied but maybe I'm missing something. 

Overall I'm not opposed to making the outputDir configurable but it seems that would hide other issues that we could fixing in another way.

**Comment by tobiasdiez** [comment]
Thanks for your quick response.

> If autocomplete doesn't work, it's definitely a bug that should be addressed either in Apollo Android or IntelliJ/Android Studio. 

I had this is VS Code (which admittedly doesn't have the best Java integration). Everything in `src/main` is included by default, other things need special configuration.

> Can you elaborate a bit more what is specific to the Java module system? 

We are still experimenting with the module system integration (see https://github.com/apollographql/apollo-android/issues/3152), but the short version is that one needs to specify the path of the module(s), see https://docs.oracle.com/javase/9/tools/jlink.htm. Thus the `build/generated` folder needs to be specified, and at the same time one needs to ensure that it is not treated as a new module.

In summary, nothing that (probably) cannot be fixed by some configuration of the external tools. It would be just easier if one could instruct the apollo plugin to generate the code in the same folder as other generated code so that this configuration has to be done only once. By the way, here is a good overview of the different conventions where to put generated code: https://stackoverflow.com/questions/22912900/location-of-generated-source-files-for-maven-directory-structure/64612308#64612308 (the build folder is not mentioned).

**Comment by martinbonnin** [comment]
Thanks for all the context! 

VSCode makes sense. I don't think they have any kind of Gradle integration so little can they know about the generated folder. 

I don't think we want to move away from `build/generated/sources/apollo` as a default. It has been serving us well so far and the same kind of layouts are used by many other libs (sqldelight, wire, android gradle plugin, ...), not to mention that it makes it easy to delete the state of a module easily (delete all build directory). 

But we could make it configurable I think. It's used [here](https://github.com/apollographql/apollo-android/blob/284aae5ee3477bdde4f56d2097b15806316ad00a/apollo-gradle-plugin/src/main/kotlin/com/apollographql/apollo/gradle/internal/ApolloPlugin.kt#L196). Maybe it's simply a matter of exposing the property to the extension.

**Comment by Siedlerchr** [comment]
This is also a problem with Eclipse: 
`Cannot nest 'jabref/build/generated/source' inside output folder 'jabref/build'`
I tried to set the outputDir in the apollo Config but that did not work:

```gradle
apollo {

  onCompilationUnit {
      outputDir.set(file("src-gen/main/java/")) // TODO not working yet
    // Overwrite some options here for single CompilationUnit if needed
  }
} 
```

**Comment by martinbonnin** [comment]
Starting with `3.0.0-alpha01`, you can set [`outputDir`](https://github.com/apollographql/apollo-android/blob/82cd2bc6c97fe092a20e67912ecdf806421f215d/apollo-gradle-plugin/src/main/kotlin/com/apollographql/apollo3/gradle/api/Service.kt#L262) in your Gradle files:

```kotlin 
apollo {
  outputDir.set("src-gen/main/java/")
}
```
